audio: copier: Fix memory corruption with TRANS_GAIN with gcc

ujfalusi opened the pull request:
he dst pointer must not be used directly while applying TRANS_GAIN to
avoid corrupting memory in heap.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>
Fixes: https://github.com/thesofproject/sof/commit/6993617c25a67cd353c71ec987b0d1f13dba39cf ("copier: gain: add processing functions")
Closes: https://github.com/thesofproject/sof/issues/10318

kv2019i reviewed (approved):
A "fixes: xxx" would be useful but not manadatory for FW commits.

ujfalusi commented:
Changes since v1:
- Fixes tag added
- Commit message reworded a bit
- dropped the reset of fade_in_sg_count as it is not needed (double checked it)

lgirdwood commented:
@lrudyX good to merge, passing on other CIs.

ujfalusi commented:
> > @lrudyX good to merge, passing on other CIs.
> 
> Failed at one test, scheduled for rerun

@lrudyX, does the test uses the copier-generic.c or the copier-hifi.c? Iow, do you compile it with GCC or Xcc?

lrudyX commented:
> > > @lrudyX good to merge, passing on other CIs.
> > 
> > 
> > Failed at one test, scheduled for rerun
> 
> @lrudyX, does the test uses the copier-generic.c or the copier-hifi.c? Iow, do you compile it with GCC or Xcc?

All tests passed after rerun. That was rather sporadic issue, not related to this PR
